### ISSUE TITLE
[MI300] Incorrect results from tl.associative_scan (cumsum / logcumsumexp) cross-wavefront reduction on gfx942

### ISSUE THREAD

**Issue by iupaikov-amd** [ROCmSoftwarePlatform/triton#947]
## Summary

With `AMDGCN_USE_BUFFER_OPS=1` (default) and `AMDGCN_USE_BUFFER_OPS=0` the result is **bit-identical** (still wrong). The scan kernel compiles to `global_load`/`global_store` in both cases, and the toggle is verified functional on a control kernel. **Disabling buffer ops does not help.** The regression is in the scan lowering (`lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp`), not the memory/indexing path.

An INNER cumulative scan (`tl.associative_scan`) over a reduction axis longer than one wavefront returns **wrong results on ~3.2% of elements** on MI300 (`gfx942`). It reproduces with a standalone `@triton.jit` kernel (the exact kernel Inductor generates for `torch.cumsum` / `torch.logcumsumexp`), with no PyTorch compiler stack involved at runtime.

The same kernel, same inputs, and same launch parameters produce **correct results on the reference Triton** and **incorrect results on current `main`**, so this is a Triton-level regression. It affects every op that lowers to a cross-wavefront scan:

- `torch.cumsum` (add combine) — inductor `test_cumsum_cuda`, `test_split_cumsum_cuda`
- `torch.logcumsumexp` (logaddexp combine) — inductor `test_logcumsumexp_cuda`

The failing pattern is a 2D `[XBLOCK, R0_BLOCK]` tile → `tl.broadcast_to` → `tl.associative_scan(..., axis=1, combine)` inside a carry loop, i.e. the cross-warp `AddPartialReduce` path in `lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp`. The `ScanOpToLLVM.cpp` diff between good/bad swaps the shared-memory load in `AddPartialReduce` to `targetInfo.loadShared(...)` and refactors `unpackInputs` to apply `removeBroadcastRegs` — both directly on this path.

Tracks nightly regression ROCm/rocActions#88.

## Environment

- **Base image:** PyTorch ROCm setup with editable PyTorch `2.11.0`; original CI container `rocm/framework-automation:triton-nightly-gha-ubuntu24.04-rocm7.2`. Triton rebuilt from the commits below on top of it.
- **GPU / arch:** AMD MI300X, `gfx942` (HIP backend, warp_size 64)
- **ROCm:** 7.14.0 (repro env)
- **Bad:** triton `3.8.0` @ `81a46fa0c04526e5df55a018ecfab72ff922f592` (`main`, 2026-06-30)
- **Good:** triton @ `ba5c1517e6f5906761cf5783036efb587026208d` (nightly reference)
- dtype `float32`, single kernel launch, `XBLOCK=1`, `R0_BLOCK=2048`, `num_warps=8`, `grid=(100,)`, reduction length `4000`

## Reproducer

Standalone script lifts the two generated `@triton.jit` scan kernels and launches them directly (via the Inductor `triton_heuristics.reduction` launcher), comparing to eager references. No `torch.compile` of a model.

Core kernel (add / cumsum variant; the logaddexp variant is identical except for the combine fn):

```python
@triton.jit
def _add(a, b):
    return a + b

@triton.jit
def cumsum_kernel(in_ptr0, out_ptr0, xnumel, r0_numel,
                  XBLOCK: tl.constexpr, R0_BLOCK: tl.constexpr):
    xnumel = 100
    r0_numel = 4000
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    x0 = xindex
    tmp3 = tl.full([XBLOCK, 1], float('nan'), tl.float32)
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK, num_stages=2):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        tmp0 = tl.load(in_ptr0 + (r0_index + 4000 * x0), r0_mask & xmask,
                       eviction_policy='evict_first', other=0.0)
        tmp2 = tl.broadcast_to(tmp0.to(tl.float32), [XBLOCK, R0_BLOCK])
        tmp4, = tl.associative_scan((tmp2,), 1, _add)
        tmp5 = triton_helpers.select_one(tmp4, r0_base == (RBLOCK - 1), dim=-1, keep_dims=True)
        tmp8 = tl.where(roffset > 0, tmp3 + tmp4, tmp4)
        tmp3 = tl.where(roffset > 0, tmp3 + tmp5, tmp5)
        tl.store(out_ptr0 + (r0_index + 4000 * x0), tmp8, r0_mask & xmask)
```

**Repro note:** a plain hand-written `tl.cumsum` does *not* reproduce, and neither does a bare `kernel[grid](...)` launch of the kernel above. Reproduction requires (a) this 2D `broadcast_to` + `associative_scan(axis=1)` pattern, (b) the config `XBLOCK=1, R0_BLOCK=2048, num_warps=8`, and (c) the `tt.divisibility=16` specialization applied by the Inductor `triton_heuristics.reduction` launcher.

## Expected behavior

Output equals the eager reference; `mismatch == 0`. This is what the reference Triton (`ba5c1517`) produces:

```
triton 3.6.0
[cumsum(add)  ] mismatch=0/400000 (0.00%) max_abs_diff=0.00109863 -> OK
[logcumsumexp ] mismatch=0/400000 (0.00%) max_abs_diff=6.67572e-06 -> OK
RESULT: ALL SCANS CORRECT
```

## Observed behavior

On triton `3.8.0` (`main` @ `81a46fa0`), ~3.2% of elements are wrong:

```
triton 3.8.0
[cumsum(add)  ] mismatch=12800/400000 (3.20%) max_abs_diff=1053.52 -> FAIL
                first bad idx=(0, 48) got=1.14321 ref=25.0449
[logcumsumexp ] mismatch=12640/400000 (3.16%) max_abs_diff=8.18053 -> FAIL
                first bad idx=(0, 48) got=0.343995 ref=4.45059
RESULT: REGRESSION REPRODUCED
```

The wrong elements cluster at wavefront boundaries within the scan axis (consistent across the `add` and `logaddexp` combines), pointing at the cross-warp partial-reduce accumulation rather than the combine function itself.

## Notes

- Reproduces deterministically for the `add` combine (bit-identical across runs); the `logaddexp` count varies by a few dozen elements run-to-run but always fails.
- Suspected first-cause: the `ScanOpToLLVM.cpp` changes on the `AddPartialReduce` / `unpackInputs` (`removeBroadcastRegs`, `targetInfo.loadShared`) path between `ba5c1517` and `81a46fa0`. A bisect over that range is the recommended next step.

### Operating System
Ubuntu 24.04

### CPU
x86_64

### GPU
AMD Instinct MI300X (gfx942)

### ROCm Version
ROCm 7.14.0

### ROCm Component
Triton (`lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp`)

### Steps to Reproduce
Build Triton from the bad/good commits above; run the standalone reproducer on a `gfx942` device.

**Comment by scxiao** [comment]
Hi @iupaikov-amd , I put together a reproducer as: https://github.com/scxiao/triton/blob/scxiao/1903_reproducer/python/tutorials/repro_scan_rocm.py. 

I tried it on both the MI350 and MI300, but cannot reproduce the issue. Do you have a complete script that can reproduce the problem? Thanks

**Comment by iupaikov-amd** [comment]
Hello @scxiao !

Sorry for the delayed response. I rechecked everything and the issue seems to only repro with buffer_ops=0. Can you please try again with the repro below?

```python
import os
os.environ.setdefault("AMDGCN_USE_BUFFER_OPS", "0")  # forces the global-load path
import sys, torch, triton, triton.language as tl
from triton.language.extra import libdevice

XBLOCK, R0_BLOCK, NUM_WARPS, XNUMEL, R0NUMEL = 1, 2048, 8, 100, 4000

@triton.jit
def _select_last(x, mask, dim: tl.constexpr):
    return tl.sum(tl.where(mask, x, x * 0.0), dim, keep_dims=True)

@triton.jit
def _add(a, b):
    return a + b

@triton.jit
def _logaddexp(a, b):
    lo = tl.minimum(a, b, tl.PropagateNan.ALL)
    hi = tl.maximum(a, b, tl.PropagateNan.ALL)
    keep = (lo != hi) | (~libdevice.isinf(lo).to(tl.int1))
    return tl.where(keep, libdevice.log1p(libdevice.exp(lo - hi)) + hi, a)

@triton.jit
def scan_kernel(in_ptr0, out_ptr0, xnumel, r0_numel,
                XBLOCK: tl.constexpr, R0_BLOCK: tl.constexpr, LOGADD: tl.constexpr):
    xnumel = 100
    r0_numel = 4000
    RBLOCK: tl.constexpr = R0_BLOCK
    xindex = tl.program_id(0) * XBLOCK + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    r0_base = tl.arange(0, R0_BLOCK)[None, :]
    tmp3 = tl.full([XBLOCK, 1], float('nan'), tl.float32)
    for r0_offset in tl.range(0, r0_numel, R0_BLOCK, num_stages=2):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        tmp0 = tl.load(in_ptr0 + (r0_index + 4000 * xindex), r0_mask & xmask,
                       eviction_policy='evict_first', other=0.0)
        tmp2 = tl.broadcast_to(tmp0.to(tl.float32), [XBLOCK, R0_BLOCK])
        if LOGADD:
            tmp4, = tl.associative_scan((tmp2,), 1, _logaddexp)
            tmp5 = _select_last(tmp4, r0_base == (RBLOCK - 1), dim=-1)
            out = tl.where(r0_offset > 0, _logaddexp(tmp3, tmp4), tmp4)
            tmp3 = tl.where(r0_offset > 0, _logaddexp(tmp3, tmp5), tmp5)
        else:
            tmp4, = tl.associative_scan((tmp2,), 1, _add)
            tmp5 = _select_last(tmp4, r0_base == (RBLOCK - 1), dim=-1)
            out = tl.where(r0_offset > 0, tmp3 + tmp4, tmp4)
            tmp3 = tl.where(r0_offset > 0, tmp3 + tmp5, tmp5)
        tl.store(out_ptr0 + (r0_index + 4000 * xindex), out, r0_mask & xmask)

def check(name, logadd, ref_fn, atol, rtol):
    torch.manual_seed(0)
    x = torch.rand((XNUMEL, R0NUMEL), device="cuda", dtype=torch.float32)
    out = torch.empty_like(x)
    scan_kernel[(XNUMEL,)](x, out, XNUMEL, R0NUMEL, XBLOCK=XBLOCK,
                           R0_BLOCK=R0_BLOCK, LOGADD=logadd, num_warps=NUM_WARPS)
    torch.cuda.synchronize()
    diff = (out - ref_fn(x)).abs()
    n = int((diff > atol + rtol * ref_fn(x).abs()).sum())
    print(f"[{name:13s}] mismatch={n}/{out.numel()} ({100*n/out.numel():.2f}%) "
          f"-> {'OK' if n == 0 else 'FAIL'}")
    return n == 0

if __name__ == "__main__":
    print("triton", triton.__version__, "|", torch.cuda.get_device_name(0))
    ok = check("cumsum(add)", False, lambda t: torch.cumsum(t, 1), 1e-5, 1e-5)
    ok &= check("logcumsumexp", True, lambda t: torch.logcumsumexp(t, 1), 1e-4, 1e-4)
    print("RESULT:", "ALL SCANS CORRECT" if ok else "REGRESSION REPRODUCED")
    sys.exit(0 if ok else 1)
```

**Comment by scxiao** [comment]
Hi @iupaikov-amd, yes, I can reproduce the issue with the above reproducer. Also, the PR: https://github.com/triton-lang/triton/pull/10739 (a new llvm bump) can fix the above issue.

**Comment by iupaikov-amd** [comment]
Confirmed fixed by triton-lang/triton#10739 ([Backend] Bump to llvm/llvm-project@850a2b1b975c).

Tested the exact PR boundary on MI300X / gfx942 (torch 2.11.0 held fixed, isolated Triton + Inductor caches per run):

| Triton | standalone scan repro | `test_cumsum` / `test_logcumsumexp` / `test_split_cumsum` |
|---|---|---|
| `cff3a73c` (parent, pre-fix) | FAIL — cumsum 3.20% / logcumsumexp 3.16% wrong (12800/400000, max_abs_diff=1053.52, first bad idx (0,48)) | FAIL / FAIL / FAIL |
| `be7e23c` (#10739 merge) | OK — 0 mismatch | PASS / PASS / PASS |
| `ba5c1517` (nightly reference) | OK — 0 mismatch | PASS / PASS / PASS |

The miscompile is present on the commit immediately before the merge and gone at the merge, so the LLVM bump in #10739 is the fix. Both the isolated scan reproducer and the three original inductor tests flip FAIL → PASS. Closing as resolved.
